Add network customization for vSphere, GCP, bare-metal installs

jboxman opened the pull request:
So, it's possible to configure the CNO for all installation types, not just AWS.

@kalexand-rh, what are you thoughts on placement for this content? Ignition configuration happens prior to creating manifests, so I placed the modules for configuring the CNO by editing manifests after. Unlike the AWS install, I notice for vSphere the example file doesn't include any networking parameters; Is that not configurable for vSphere?

This probably needs deeper investigation but I wanted to start this.

Thanks!

kalexand-rh commented:
@jboxman, that sounds reasonable, and I can take a closer look when you're closer to merging.

@dav1x, do you have opinions about adding networking parameters for vSphere?
> Unlike the AWS install, I notice for vSphere the example file doesn't include any networking parameters; Is that not configurable for vSphere?

jboxman commented:
@kalexand-rh, I've updated this. The big outstanding question is what `networking` parameters, if any, are available in the `install-config.yaml` for bare-metal and vSphere respectively. Based on that, I'll need to customize or not include the networking parameters table in each assembly.

kalexand-rh commented:
@jboxman, all of the parameters that I am aware of are in the `install-config.yaml` examples already. You might want to check with your devs.

jboxman commented:
As near as I can tell, the `Networking` parameter is always acceptable; where or not it has any effect for a given platform I don't know so far.

https://github.com/openshift/installer/blob/4edffaead4e6f4d4e539daa1cf7ea3444fe69c49/pkg/types/installconfig.go#L56

jboxman commented:
@kalexand-rh, I've updated the PR and fixed the build error. The previews are here:

- [bare-metal](https://inst-nw-customization-vsphere-baremetal--ocpdocs.netlify.com/openshift-enterprise/latest/installing/installing_bare_metal/installing-bare-metal-network-customizations.html)
- [vSphere](https://inst-nw-customization-vsphere-baremetal--ocpdocs.netlify.com/openshift-enterprise/latest/installing/installing_vsphere/installing-vsphere-network-customizations.html)

kalexand-rh commented:
@jboxman, it seems reasonable to me and is consistent with our other approach to this configuration. @abhinavdahiya, is there any other context that you think might be helpful in helping users know when they'd want to use the network operator in a bare metal or vSphere installation instead of using the standard options?

jboxman commented:
@kalexand-rh, what's the next step for this PR? Thanks!

kalexand-rh commented:
@jboxman, I'd probably push for a deeper dev review again before sending it to QE for verification.

jboxman commented:
@kalexand-rh, not sure who to ask from engineering as this cuts across concerns.

@squeed @danwinship - The CNO can be customized during install for all supported platforms  in 4.2, yes?

Thanks.

squeed commented:
Correct - the CNO is platform independent. If possible, it would be nice to see it removed from the platform-specific documentation.

danwinship commented:
Minor clarification: `networking.clusterNetwork`, `networking.serviceNetwork`, and `networking.networkType` are CNO configuration and are platform-independent (although `Kuryr` is only a valid value for `networkType` when using OpenStack). `networking.machineCIDR` is not CNO config, and is not allowed/effective on some platforms.

dav1x commented:
@kalexand-rh Sorry for the late response. I think the nsx-t plugin would be the SDN plugin specific to vSphere. But, from my understanding they are working on an operator for installation and configuration of NSX-T.  I'm not really an NSX expert unfortunately.

jboxman commented:
@anuragthehatter, this PR adds installation network customization for GCP, vSphere, and bare metal. Can you confirm whether the content is accurate or suggest someone who can?

Thanks!

Previews:

- [GCP](http://file.bos.redhat.com/jboxman/20191106/installing/installing_gcp/installing-gcp-network-customizations.html)
- [vSphere](http://file.bos.redhat.com/jboxman/20191106/installing/installing_vsphere/installing-vsphere-network-customizations.html)
- [Bare metal](http://file.bos.redhat.com/jboxman/20191106/installing/installing_bare_metal/installing-bare-metal-network-customizations.html)

jboxman commented:
@kalexand-rh, latest iteration of this, which adds GCP. Once this is merged, I can remove the conditionals and TP notice in a 4.3-only PR for OVN GA in 4.3 on (hopefully) all platforms.

anuragthehatter commented:
Looks good to me. Thanks for incorporating it.
/lgtm

jboxman commented:
@anuragthehatter, thanks!

jboxman commented:
/cherry-pick enterprise-4.3

jboxman commented:
/cherry-pick enterprise-4.2
